pyro/infer/mcmc/adaptation.py

```diff
@@ -350,3 +352,247 @@ def unscale(self, r):
             r_flat = torch.cat([r[site_name].reshape(-1) for site_name in site_names])
             u[site_names] = _matmul(mass_matrix_sqrt_inverse, r_flat)
         return u
+
+
+def _arrowhead_sqrt(x):
+    assert isinstance(x, ArrowheadMatrix)
+    head_size = x.top.size(0)
+    if head_size == 0:
+        return ArrowheadUpperSqrtMatrix(x.top, x.bottom_diag.sqrt())
+
+    A, B = x.top[:, :head_size], x.top[:, head_size:]
+    # NB: the complexity is O(N * head_size^2)
+    # ref: https://en.wikipedia.org/wiki/Schur_complement#Background
+    Dsqrt = x.bottom_diag.sqrt()
+    B_Dsqrt = B / Dsqrt.unsqueeze(-2)  # shape: head_size x N
+    schur_complement = A - B_Dsqrt.matmul(B_Dsqrt.t())  # complexity: head_size^2 x N
+    # we will decompose schur_complement to U @ U.T (so that the sqrt matrix
+    # is upper triangular) using some `flip` operators:
+    #   flip(cholesky(flip(schur_complement)))
+    top_left = torch.flip(torch.flip(schur_complement, (-2, -1)).cholesky(), (-2, -1))
```

Comment:
this isn't important but why are these flips necessary? it seems like you should be able to avoid them....

Reply:
You're right that those flips aren't necessary. But having an upper triangular form is convenient for testing. This also makes those sqrt forms consistent with BlockMassMatrix (its sqrt/sqrt_inverse is also upper triangular). And we just flip several times (at the end of each adaptation windows) during adaptation, so their effects are negligible.